src/System.Net.HttpListener/tests/HttpListenerContextTests.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net.Sockets;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Tests
+{
+    public class HttpListenerContextTests : IDisposable
+    {
+        private HttpListenerFactory Factory { get; }
+        private ClientWebSocket Socket { get; }
+
+        public HttpListenerContextTests()
+        {
+            Factory = new HttpListenerFactory();
+            Socket = new ClientWebSocket();
+        }
+
+        public void Dispose()
+        {
+            Factory.Dispose();
+            Socket.Dispose();
+        }
+
+        public static bool IsNotWindows7OrUapCore { get; } = !PlatformDetection.IsWindows7 && PlatformDetection.IsNotOneCoreUAP;
+
+        public static IEnumerable<object[]> SubProtocol_TestData()
+        {
+            yield return new object[] { new string[0], null };
+
+            yield return new object[] { new string[] { "MyProtocol1" }, null };
+            yield return new object[] { new string[] { "MyProtocol" }, "MyProtocol" };
+            yield return new object[] { new string[] { "MyProtocol" }, "myPROTOCOL" };
+
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, null };
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, "MyProtocol2" };
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7OrUapCore))]
+        [MemberData(nameof(SubProtocol_TestData))]
+        public async Task AcceptWebSocketAsync_ValidSubProtocol_Success(string[] clientProtocols, string serverProtocol)
+        {
+            HttpListenerContext context = await GetWebSocketContext(clientProtocols);
+            HttpListenerWebSocketContext socketContext = await context.AcceptWebSocketAsync(serverProtocol);
+            Assert.Equal(serverProtocol, socketContext.WebSocket.SubProtocol);
+        }
+
+        [ConditionalFact(nameof(IsNotWindows7OrUapCore))]
+        // The managed implementation doesn't validate that the socket is actually a web socket.
+        // This means that HttpListener can connect to a Socket masquerading as a web socket.
+        [ActiveIssue(18128, TestPlatforms.AnyUnix)]
+        public async Task AcceptWebSocketAsync_SocketSpoofingAsWebSocket_ThrowsWebSocketException()
+        {
+            await GetSocketContext(new string[] { "Connection: Upgrade", "Upgrade: websocket", "Sec-WebSocket-Version: 13", "Sec-WebSocket-Key: Key" }, context =>
+            {
+                Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(null)).Wait();
```

Comment:
Blocking like this with Wait is problematic.  The problem is that xunit uses a synchronization context that only allows for N pieces of test code to be running at the same time.  Let's say N == 1.  If that test blocks waiting for an async operation to complete, and that async operation itself has queued some work back to the SynchronizationContext to be run (the default with `await task;`), then the only thread that could possibly execute that continuation is blocked waiting for that continuation, and you deadlock.

The right pattern here is to change GetSocketContext to take a `Func<context, Task>` instead of an `Action<context>`.  Then you'd do:
```C#
async context =>
{
    await Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(null));
]
```